core/src/main/java/org/bitcoinj/core/StoredUndoableBlock.java

```diff
@@ -27,14 +28,18 @@
  * transactions (if the inputs for the block have not been tested to work)
  * or the set of transaction outputs created/destroyed when the block is
  * connected.
+ * <p>
+ * Instances of this class are immutable.
  */
-public class StoredUndoableBlock {
+public final class StoredUndoableBlock {
     
-    Sha256Hash blockHash;
+    private final Sha256Hash blockHash;
     
     // Only one of either txOutChanges or transactions will be set
-    private TransactionOutputChanges txOutChanges;
-    private List<Transaction> transactions;
+    @Nullable
+    private final TransactionOutputChanges txOutChanges;
+    @Nullable
+    private final List<Transaction> transactions;
     
     public StoredUndoableBlock(Sha256Hash hash, TransactionOutputChanges txOutChanges) {
```

Comment:
`txOutChanges` should be annotated as nullable or non-nullable as well as should `transactions` in the other constructor. Or, if we are trying to consistently have unannotated parameters be considered non-nullable, we should at least add some JavaDoc to the two constructors explaining their use-cases.

Reply:
We could also consider a private canonical constructor with a pre-condition that enforces that both can't be null or both can't be set.

Reply:
I would add JavaDocs to the constructors if I would understand the usecases. I'm currently happy we're able to make this class shallowly immutable without any tests failing.

Reply:
The private canonical constructor I think is a candidate for when we migrate the existing constructors to static constructors. But that's out of scope I'd say.